File: backend/server/src/routes/ws.rs
use crate::protocol::{ClientToServerMessage, RoomPeer, ServerToClientMessage};
use crate::session::require_session;
use crate::state::AppState;
use crate::store::message_store::normalize_target_id;
use crate::utils::sanitize_room_id;
use futures_util::{SinkExt, StreamExt};
use salvo::prelude::*;
use salvo::websocket::{Message, WebSocketUpgrade};
use tokio::sync::mpsc;
use tokio_stream::wrappers::UnboundedReceiverStream;

#[handler]
pub async fn room_ws(
    req: &mut Request,
    depot: &mut Depot,
    res: &mut Response,
) -> Result<(), StatusError> {
    let state = depot
        .obtain::<AppState>()
        .map_err(|_| StatusError::internal_server_error())?
        .clone();
    let room_id = sanitize_room_id(&req.param::<String>("room_id").unwrap_or_default());
    if room_id.trim().is_empty() {
        return Err(StatusError::bad_request().brief("missing room_id"));
    }

    let session = require_session(req, &state.config.session_secret)
        .map_err(|error| {
            StatusError::internal_server_error().brief(format!("session decode error: {error}"))
        })?
        .ok_or_else(|| StatusError::unauthorized().brief("missing session"))?;

    let client_id = session.clientId.clone();
    let nickname = session.nickname.clone();
    WebSocketUpgrade::new()
        .upgrade(req, res, move |ws| async move {
            handle_socket(state, room_id, client_id, nickname, ws).await;
        })
        .await
}

async fn handle_socket(
    state: AppState,
    room_id: String,
    client_id: String,
    nickname: String,
    ws: salvo::websocket::WebSocket,
) {
    let (mut ws_tx, mut ws_rx) = ws.split();
    let (tx, rx) = mpsc::unbounded_channel::<String>();
    tokio::spawn(async move {
        let mut rx = UnboundedReceiverStream::new(rx);
        while let Some(message) = rx.next().await {
            if ws_tx.send(Message::text(message)).await.is_err() {
                break;
            }
        }
    });

    let joined = state
        .room_hub
        .join_room(&room_id, &client_id, &nickname, tx.clone())
        .await;
    let connection_id = joined.connection_id.clone();
    let messages = match state
        .message_store
        .list_visible_messages(&room_id, &client_id)
        .await
    {
        Ok(messages) => messages,
        Err(error) => {
            tracing::error!(room_id, client_id, error = %error, "failed to load visible messages");
            Vec::new()
        }
    };
    let snapshot = ServerToClientMessage::RoomSnapshot {
        roomId: room_id.clone(),
        peers: joined.peers,
        messages,
        serverTime: crate::utils::now_ms(),
    };
    let _ = state.room_hub.send_json(&tx, &snapshot);
    state
        .room_hub
        .broadcast(
            &room_id,
            Some(&client_id),
            None,
            &ServerToClientMessage::PeerJoined {
                peer: RoomPeer {
                    clientId: joined.joined_peer.clientId.clone(),
                    nickname: joined.joined_peer.nickname.clone(),
                    joinedAt: joined.joined_peer.joinedAt,
                },
            },
        )
        .await;

    while let Some(Ok(message)) = ws_rx.next().await {
        if message.is_close() {
            break;
        }
        if !message.is_text() {
            continue;
        }

        let Some(text) = message.as_str().ok().map(ToOwned::to_owned) else {
            continue;
        };
        if !state
            .room_hub
            .is_current_connection(&room_id, &client_id, &connection_id)
            .await
        {
            break;
        }
        state
            .room_hub
            .touch(&room_id, &client_id, &connection_id)
            .await;

        let payload: ClientToServerMessage = match serde_json::from_str(&text) {
            Ok(payload) => payload,
            Err(_) => {
                let _ = state.room_hub.send_json(
                    &tx,
                    &ServerToClientMessage::Error {
                        code: "invalid_json".to_owned(),
                        message: "消息解析失败。".to_owned(),
                    },
                );
                continue;
            }
        };

        match payload {
            ClientToServerMessage::Ping => {
                let _ = state.room_hub.send_json(
                    &tx,
                    &ServerToClientMessage::Pong {
                        serverTime: crate::utils::now_ms(),
                    },
                );
            }
            ClientToServerMessage::SetProfile { nickname } => {
                rename_peer(&state, &room_id, &client_id, &connection_id, nickname).await;
            }
            ClientToServerMessage::ChatSend { text, targetId } => {
                handle_chat_send(&state, &room_id, &client_id, &connection_id, text, targetId)
                    .await;
            }
            ClientToServerMessage::Signal { targetId, payload } => {
                forward_signal(
                    &state,
                    &room_id,
                    &client_id,
                    &connection_id,
                    &targetId,
                    payload,
                )
                .await;
            }
            ClientToServerMessage::RelayFileAnnounced { file } => {
                handle_relay_file(&state, &room_id, &client_id, &connection_id, file, &tx).await;
            }
        }
    }

    remove_peer(&state, &room_id, &client_id, &connection_id).await;
}

async fn rename_peer(
    state: &AppState,
    room_id: &str,
    client_id: &str,
    connection_id: &str,
    nickname: String,
) {
    let Some(updated) = state
        .room_hub
        .rename_peer(room_id, client_id, connection_id, nickname)
        .await
    else {
        return;
    };

    state
        .room_hub
        .broadcast(
            room_id,
            None,
            None,
            &ServerToClientMessage::PeerJoined { peer: updated },
        )
        .await;
}

async fn handle_chat_send(
    state: &AppState,
    room_id: &str,
    client_id: &str,
    connection_id: &str,
    text: String,
    target_id: Option<String>,
) {
    if text.trim().is_empty() {
        return;
    }
    if !state
        .room_hub
        .is_current_connection(room_id, client_id, connection_id)
        .await
    {
        return;
    }
    let Some(from_name) = state.room_hub.display_name_for(room_id, client_id).await else {
        return;
    };
    let normalized_target = match normalize_target_id(client_id, target_id) {
        Some(target) if state.room_hub.is_client_connected(room_id, &target).await => Some(target),
        _ => None,
    };
    let recipients = normalized_target
        .clone()
        .map(|target| vec![client_id.to_owned(), target]);
    let message = match state
        .message_store
        .persist_text_message(room_id, client_id, &from_name, normalized_target, &text)
        .await
    {
        Ok(message) => message,
        Err(error) => {
            tracing::error!(room_id, client_id, error = %error, "failed to persist text message");
            return;
        }
    };

    state
        .room_hub
        .broadcast(
            room_id,
            None,
            recipients.as_deref(),
            &ServerToClientMessage::ChatEvent { message },
        )
        .await;
}

async fn forward_signal(
    state: &AppState,
    room_id: &str,
    from_id: &str,
    connection_id: &str,
    target_id: &str,
    payload: crate::protocol::SignalEnvelope,
) {
    if !state
        .room_hub
        .is_current_connection(room_id, from_id, connection_id)
        .await
    {
        return;
    }
    let target = state
        .room_hub
        .resolve_signal_target(room_id, from_id, connection_id, target_id)
        .await;

    if let Some(target) = target {
        let _ = state.room_hub.send_json(
            &target,
            &ServerToClientMessage::Signal {
                fromId: from_id.to_owned(),
                payload,
            },
        );
    }
}

async fn handle_relay_file(
    state: &AppState,
    room_id: &str,
    client_id: &str,
    connection_id: &str,
    file: crate::protocol::RelayFileAnnouncement,
    tx: &crate::state::ClientTx,
) {
    if !state
        .room_hub
        .is_current_connection(room_id, client_id, connection_id)
        .await
    {
        return;
    }
    let Some(from_name) = state.room_hub.display_name_for(room_id, client_id).await else {
        return;
    };
    let confirmed_file = match state
        .relay_store
        .confirm_announced_file(room_id, client_id, file)
        .await
    {
        Ok(file) => file,
        Err(error) => {
            let _ = state.room_hub.send_json(
                tx,
                &ServerToClientMessage::Error {
                    code: "invalid_relay_file".to_owned(),
                    message: format!("relay 文件确认失败: {error}"),
                },
            );
            return;
        }
    };
    let normalized_target = match normalize_target_id(client_id, confirmed_file.targetId.clone()) {
        Some(target) if state.room_hub.is_client_connected(room_id, &target).await => Some(target),
        _ => None,
    };
    let recipients = normalized_target
        .clone()
        .map(|target| vec![client_id.to_owned(), target]);
    let message = match state
        .message_store
        .persist_relay_file_message(
            room_id,
            client_id,
            &from_name,
            normalized_target,
            confirmed_file,
        )
        .await
    {
        Ok(message) => message,
        Err(error) => {
            let _ = state.room_hub.send_json(
                tx,
                &ServerToClientMessage::Error {
                    code: "relay_persist_failed".to_owned(),
                    message: format!("relay 文件写库失败: {error}"),
                },
            );
            return;
        }
    };

    state
        .room_hub
        .broadcast(
            room_id,
            None,
            recipients.as_deref(),
            &ServerToClientMessage::ChatEvent { message },
        )
        .await;
}

async fn remove_peer(state: &AppState, room_id: &str, client_id: &str, connection_id: &str) {
    let removed = state
        .room_hub
        .leave_room(room_id, client_id, connection_id)
        .await;

    if removed {
        state
            .room_hub
            .broadcast(
                room_id,
                Some(client_id),
                None,
                &ServerToClientMessage::PeerLeft {
                    clientId: client_id.to_owned(),
                },
            )
            .await;
    }
}

File: backend/server/internal/httpapi/ws.go
package httpapi

import (
	"context"
	"encoding/json"
	"log/slog"
	"net/http"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/gorilla/websocket"

	"github.com/baisuipingan/patrick-im/backend/server/internal/messages"
	"github.com/baisuipingan/patrick-im/backend/server/internal/protocol"
	"github.com/baisuipingan/patrick-im/backend/server/internal/realtime"
	"github.com/baisuipingan/patrick-im/backend/server/internal/session"
	"github.com/baisuipingan/patrick-im/backend/server/internal/util"
)

const (
	writeWait      = 10 * time.Second
	pongWait       = 60 * time.Second
	pingPeriod     = (pongWait * 9) / 10
	maxMessageSize = protocol.MaxChatTextBytes + 128*1024
)

const (
	wsProtocolName          = "patrick-im"
	wsSessionProtocolPrefix = "patrick-im-session."
)

var upgrader = websocket.Upgrader{
	ReadBufferSize:  4096,
	WriteBufferSize: 4096,
	CheckOrigin: func(r *http.Request) bool {
		return true
	},
}

func (api *API) roomWS(c *gin.Context) {
	roomID := util.SanitizeRoomID(c.Param("room_id"))
	if strings.TrimSpace(roomID) == "" {
		api.fail(c, http.StatusBadRequest, "missing room_id")
		return
	}
	sess, err := api.requireWebSocketSession(c.Request)
	if err != nil {
		api.fail(c, http.StatusUnauthorized, "missing session")
		return
	}
	responseHeader := http.Header{}
	if websocketSubprotocolRequested(c.Request, wsProtocolName) {
		responseHeader.Set("Sec-WebSocket-Protocol", wsProtocolName)
	}
	conn, err := upgrader.Upgrade(c.Writer, c.Request, responseHeader)
	if err != nil {
		api.logger.Warn("websocket upgrade failed", "error", err)
		return
	}
	nickname := sess.Nickname
	if queryName := strings.TrimSpace(c.Query("nickname")); queryName != "" {
		nickname = queryName
	}
	go api.handleSocket(roomID, sess.ClientID, nickname, conn)
}

func (api *API) requireWebSocketSession(r *http.Request) (session.Payload, error) {
	if payload, err := session.Require(r, api.cfg.SessionSecret); err == nil {
		return payload, nil
	}
	token := websocketSessionToken(r)
	if token == "" {
		return session.Payload{}, http.ErrNoCookie
	}
	payload, err := session.ReadToken(token, api.cfg.SessionSecret)
	if err != nil || payload == nil {
		if err != nil {
			return session.Payload{}, err
		}
		return session.Payload{}, http.ErrNoCookie
	}
	return *payload, nil
}

func websocketSessionToken(r *http.Request) string {
	for _, item := range websocketSubprotocols(r) {
		if strings.HasPrefix(item, wsSessionProtocolPrefix) {
			return strings.TrimPrefix(item, wsSessionProtocolPrefix)
		}
	}
	return ""
}

func websocketSubprotocolRequested(r *http.Request, protocol string) bool {
	for _, item := range websocketSubprotocols(r) {
		if item == protocol {
			return true
		}
	}
	return false
}

func websocketSubprotocols(r *http.Request) []string {
	raw := r.Header.Values("Sec-WebSocket-Protocol")
	out := make([]string, 0)
	for _, header := range raw {
		for _, item := range strings.Split(header, ",") {
			item = strings.TrimSpace(item)
			if item != "" {
				out = append(out, item)
			}
		}
	}
	return out
}

func (api *API) handleSocket(roomID, clientID, nickname string, conn *websocket.Conn) {
	tx := realtime.NewClientTx()
	joined := api.hub.JoinRoom(roomID, clientID, nickname, tx)
	if joined.ReplacedTx != nil {
		joined.ReplacedTx.Close()
	}

	go writePump(conn, tx, api.logger)
	api.sendRoomSnapshot(roomID, clientID, tx, joined.Peers)
	api.hub.Broadcast(roomID, &clientID, nil, protocol.ServerToClientMessage{
		Type: "peer-joined",
		Peer: &joined.JoinedPeer,
	})

	readDone := make(chan struct{})
	go func() {
		defer close(readDone)
		api.readPump(roomID, clientID, joined.ConnectionID, tx, conn)
	}()
	<-readDone
	tx.Close()
	_ = conn.Close()
	api.removePeer(roomID, clientID, joined.ConnectionID)
}

func (api *API) readPump(roomID, clientID, connectionID string, tx *realtime.ClientTx, conn *websocket.Conn) {
	conn.SetReadLimit(maxMessageSize)
	_ = conn.SetReadDeadline(time.Now().Add(pongWait))
	conn.SetPongHandler(func(string) error {
		return conn.SetReadDeadline(time.Now().Add(pongWait))
	})
	for {
		messageType, payload, err := conn.ReadMessage()
		if err != nil {
			return
		}
		if messageType != websocket.TextMessage {
			continue
		}
		if !api.hub.IsCurrentConnection(roomID, clientID, connectionID) {
			return
		}
		api.hub.Touch(roomID, clientID, connectionID)
		var message protocol.ClientToServerMessage
		if err := json.Unmarshal(payload, &message); err != nil {
			api.hub.SendJSON(tx, protocol.ServerToClientMessage{
				Type:         "error",
				Code:         "invalid_json",
				ErrorMessage: "消息解析失败。",
			})
			continue
		}
		api.handleClientMessage(roomID, clientID, connectionID, tx, message)
	}
}

func writePump(conn *websocket.Conn, tx *realtime.ClientTx, logger *slog.Logger) {
	ticker := time.NewTicker(pingPeriod)
	defer func() {
		ticker.Stop()
		_ = conn.Close()
	}()
	for {
		select {
		case payload, ok := <-tx.C():
			_ = conn.SetWriteDeadline(time.Now().Add(writeWait))
			if !ok {
				_ = conn.WriteMessage(websocket.CloseMessage, []byte{})
				return
			}
			if err := conn.WriteMessage(websocket.TextMessage, payload); err != nil {
				logger.Warn("websocket write failed", "error", err)
				return
			}
		case <-ticker.C:
			_ = conn.SetWriteDeadline(time.Now().Add(writeWait))
			if err := conn.WriteMessage(websocket.PingMessage, nil); err != nil {
				return
			}
		}
	}
}

func (api *API) sendRoomSnapshot(roomID, clientID string, tx *realtime.ClientTx, peers []protocol.RoomPeer) {
	history, err := api.messages.ListVisibleMessages(context.Background(), roomID, clientID)
	if err != nil {
		api.logger.Error("load visible messages failed", "room_id", roomID, "client_id", clientID, "error", err)
		history = []protocol.ChatMessage{}
	}
	api.hub.SendJSON(tx, protocol.ServerToClientMessage{
		Type:       "room-snapshot",
		RoomID:     roomID,
		Peers:      peers,
		Messages:   history,
		ServerTime: util.NowMS(),
	})
}

func (api *API) removePeer(roomID, clientID, connectionID string) {
	removed := api.hub.LeaveRoom(roomID, clientID, connectionID)
	if removed {
		api.hub.Broadcast(roomID, &clientID, nil, protocol.ServerToClientMessage{
			Type:     "peer-left",
			ClientID: clientID,
		})
	}
}

func (api *API) handleClientMessage(roomID, clientID, connectionID string, tx *realtime.ClientTx, payload protocol.ClientToServerMessage) {
	switch payload.Type {
	case "ping":
		api.hub.SendJSON(tx, protocol.ServerToClientMessage{Type: "pong", ServerTime: util.NowMS()})
	case "set-profile":
		api.renamePeer(roomID, clientID, connectionID, payload.Nickname)
	case "chat-send":
		api.handleChatSend(roomID, clientID, connectionID, tx, payload.Text, optionalString(payload.TargetID))
	case "signal":
		if payload.Payload != nil {
			api.forwardSignal(roomID, clientID, connectionID, payload.TargetID, *payload.Payload)
		}
	case "relay-file-announced":
		if payload.File != nil {
			api.handleRelayFile(roomID, clientID, connectionID, *payload.File, tx)
		}
	}
}

func (api *API) renamePeer(roomID, clientID, connectionID, nickname string) {
	peer, ok := api.hub.RenamePeer(roomID, clientID, connectionID, nickname)
	if !ok {
		return
	}
	api.hub.Broadcast(roomID, nil, nil, protocol.ServerToClientMessage{
		Type: "peer-joined",
		Peer: peer,
	})
}

func (api *API) handleChatSend(roomID, clientID, connectionID string, tx *realtime.ClientTx, text string, targetID *string) {
	text = strings.TrimSpace(text)
	if text == "" {
		return
	}
	if len([]byte(text)) > protocol.MaxChatTextBytes {
		api.hub.SendJSON(tx, protocol.ServerToClientMessage{
			Type:         "error",
			Code:         "chat_text_too_large",
			ErrorMessage: "文字内容太长，请拆成多条发送或作为文件发送。",
		})
		return
	}
	if !api.hub.IsCurrentConnection(roomID, clientID, connectionID) {
		return
	}
	fromName, ok := api.hub.DisplayNameFor(roomID, clientID)
	if !ok {
		return
	}
	normalizedTarget := messages.NormalizeTargetID(clientID, targetID)
	if normalizedTarget != nil && !api.hub.IsClientConnected(roomID, *normalizedTarget) {
		normalizedTarget = nil
	}
	recipients := recipientsFor(normalizedTarget, clientID)
	message, err := api.messages.PersistTextMessage(context.Background(), roomID, clientID, fromName, normalizedTarget, text)
	if err != nil {
		api.logger.Error("persist text message failed", "room_id", roomID, "client_id", clientID, "error", err)
		api.hub.SendJSON(tx, protocol.ServerToClientMessage{
			Type:         "error",
			Code:         "chat_text_persist_failed",
			ErrorMessage: "文字发送失败，请稍后重试。",
		})
		return
	}
	api.hub.Broadcast(roomID, nil, recipients, protocol.ServerToClientMessage{
		Type:    "chat-event",
		Message: &message,
	})
}

func (api *API) forwardSignal(roomID, fromID, connectionID, targetID string, payload protocol.SignalEnvelope) {
	if !api.hub.IsCurrentConnection(roomID, fromID, connectionID) {
		return
	}
	target := api.hub.ResolveSignalTarget(roomID, fromID, connectionID, targetID)
	if target == nil {
		return
	}
	api.hub.SendJSON(target, protocol.ServerToClientMessage{
		Type:    "signal",
		FromID:  fromID,
		Payload: &payload,
	})
}

func (api *API) handleRelayFile(roomID, clientID, connectionID string, file protocol.RelayFileAnnouncement, tx *realtime.ClientTx) {
	if !api.hub.IsCurrentConnection(roomID, clientID, connectionID) {
		return
	}
	fromName, ok := api.hub.DisplayNameFor(roomID, clientID)
	if !ok {
		return
	}
	normalizedTarget := messages.NormalizeTargetID(clientID, file.TargetID)
	if normalizedTarget != nil && !api.hub.IsClientConnected(roomID, *normalizedTarget) {
		normalizedTarget = nil
	}
	recipients := recipientsFor(normalizedTarget, clientID)
	outcome, err := api.messages.PersistConfirmedRelayFileMessage(context.Background(), roomID, clientID, fromName, normalizedTarget, file)
	if err != nil {
		api.hub.SendJSON(tx, protocol.ServerToClientMessage{
			Type:         "error",
			Code:         "relay_file_announce_failed",
			ErrorMessage: "relay 文件确认失败。",
		})
		return
	}
	event := protocol.ServerToClientMessage{Type: "chat-event", Message: &outcome.Message}
	if outcome.Created {
		api.hub.Broadcast(roomID, nil, recipients, event)
		return
	}
	api.hub.SendJSON(tx, event)
}
